### ISSUE TITLE
[BUG] ProgressReceiver reporting upload finished before upload has finished

### ISSUE THREAD

**Issue by MtJand** [Azure/azure-sdk-for-java#7985]
**Describe the bug**
When uploading a file to blob storage using BlobAsyncClient and ProgressReceiver to monitor progress it is reported that all the bytes have been transferred very quickly as the upload starts. The outcome is the same both when a ProgressReciever is applied to the Flux using ProgressReporter.addProgressReporting() and when a ProgressReceiver is passed as an argument to ParallelTransferOptions which is then Passed to BlobAsyncClient.uploadWithResponse().

***Exception or Stack Trace***
N/A

**To Reproduce**
Upload a file using BlobAsyncClient.uploadWithResponse() and attempt to monitor progress using a ProgressReceiver

***Code Snippet***
`
                    Flux<ByteBuffer> data = ByteBufFlux.fromPath(Paths.get("Crowd.mov")).asByteBuffer();

                    ParallelTransferOptions parallelTransferOptions = new ParallelTransferOptions(16384,
                                                                                                  null,
                                                                                                  bytesTransferred -> 
                                                                                                          System.out.printf("Bytes transferred: %s", bytesTransferred));

                    Mono<Response<BlockBlobItem>> upload = blobAsyncClient.uploadWithResponse(data,
                                                                                         parallelTransferOptions,
                                                                                         null,
                                                                                         null,
                                                                                         AccessTier.HOT,
                                                                                         null);

                    upload.subscribe(response -> System.out.printf("Uploaded BlockBlob"));`

**Expected behavior**
ProgressReceiver to report the bytes transferred over the network to monitor progress of the upload. 

**Screenshots**
N/A

**Setup (please complete the following information):**
 - OS: Win10
 - IDE :IntelliJ
 - Version of the Library used: 12.3.0

**Additional context**
N/A

**Information Checklist**
Kindly make sure that you have added all the following information above and checkoff the required fields otherwise we will treat the issuer as an incomplete report
- [/] Bug Description Added
- [/] Repro Steps Added
- [/] Setup information Added

**Comment by joshfree** [comment]
Thanks for reporting this issue @MtJand. @rickle-msft @gapra-msft could you please take a look?

**Comment by joshfree** [comment]
Related https://github.com/Azure/azure-sdk-for-java/issues/7912

**Comment by rickle-msft** [comment]
Can you share the size of your file and the level of parallelism? We are reporting the number of bytes as they are emitted from your flux (i.e. being put on the wire), and if your file is small enough, we may issue all the stage block requests at once and hence rather quickly report that the file is uploaded. I suspect you want the update only after we get a response from the service that the staged data has been accepted? 

I'll need to check with my team on whether this is the desired behavior or whether we will recommend that you do something like distinguish between the data being staged and the data being committed when you track progress. This might look like only showing 99% overall progress even if all the data has been sent until the entire operation including the commit is done.

A few dev notes:
We could possibly achieve delayed reporting in uploads by doing something like 
```
stageMono = stageBlock().cache() // so we don’t double subscribe
addParallelProgressReporting(stageMono.map(response->dataSize),…) // Requires adding an overload to accept Publisher<Integer>. This just reports progress upon emission of the Integer. This ensures we only report the update after the response has been returned. 
return stageMono.
```

**Comment by rickle-msft** [comment]
I also remember seeing in an email thread that you guys had a concern about copy progress reporting? Is that still an issue for you? It does look like the CopyInfo object is missing that type, and I can create an issue to track adding it if you would like.

**Comment by MtJand** [comment]
The original issue report regarding upload progress was with a 30MB file but uploading larger files will report a portion of bytes transmitted straight away and then pause (I assume whilst the original block is uploaded) and then repeat with the next portion of bytes. 

I suspected the ProgressReceiver was monitoring the data being emitted from the Flux but the docs weren't very clear on this. It is the progress after the staged data has been committed that we need. 

In terms of the copy progress reporting: it would be ideal if we could monitor the progress but it's not a huge concern as they complete very quickly.

**Comment by rickle-msft** [comment]
Thank you for your response! Apologies for the long response but there's a TL;DR down below if needed. 

Forgive me if this feels overly pedantic, but I need to clarify a detail of your response before moving forward. You mentioned that you need the progress after the staged data has been _committed_. Uploading to Storage has two parts. We upload the data to the service by staging as many blocks as necessary. As the last request of the operation, we call a `commitBlockList`, which commits all the blocks at once. It is only after this commit that the data is finalized and available for read or other operations. 

Given this definition of committing, if you are looking for progress as data is committed, you are always going to see the committed data going from 0 to 100 all at once. If you are looking for reports of when blocks of data have been successfully accepted by the service as staged, but not committed, that is probably doable pending a design discussion on our team. I'm assuming you prefer the latter, but want to be extra clear (since our docs weren't :) ). Can you also share why this change is important to your scenario?

The reason I ask is that if we do make this change, you are likely to see a similar reporting pattern, just with an initial latency. Our upload is fairly burst-y in that we tend to make as many parallel requests as we are allowed, and then we are forced to wait until those complete before issuing more. Since they all started at the same time and are generally the same size except the last block, they will typically all complete at the same time and more requests will be issued from there, likely in quick succession and looking as though they were another "batch" but not actually correlated to each other.

TL;DR When you say you're looking for progress as data is committed, do you actually mean committed as defined by the service or do you mean data has been accepted by the service as a staged block? And given that the reporting pattern will be similar but with an initial latency, can you please share why this change is necessary for your scenario?

**Comment by MtJand** [comment]
Thanks for clearing that up. 

I did mean as the data was accepted as staged, not committed. My apologies for any confusion caused. However, now that I understand how the upload process works I can't see much benefit in this change anymore and with some tweaking of the block size frequent progress updates can be received for larger files. 

Closing this issue.
